### PR TITLE
auth-backend: refactor providers to accept options, and delay profile transforms

### PR DESCRIPTION
**Pull request by Rugvip**
## Hey, I just made a Pull Request!

Some preparation for more customizable auth providers, no actual changes to behavior (I hope :grin: )

See #4532 for what this enables

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

### CODE REVIEW

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4531-t1xgxaaebp3rdur12d5jxfrf73luuly1.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/602a610065f8172627f09bf1

File: plugins/auth-backend/src/providers/auth0/provider.ts [hunk 1]
```diff
@@ -61,20 +62,16 @@ export class Auth0AuthProvider implements OAuthHandlers {
         accessToken: any,
         refreshToken: any,
         params: any,
-        rawProfile: passport.Profile,
-        done: PassportDoneCallback<OAuthResponse, PrivateInfo>,
+        fullProfile: passport.Profile,
```
Comment (freben) [inline plugins/auth-backend/src/providers/auth0/provider.ts:65]: Would it not make sense to keep calling these `rawProfile` (as above) or `passportProfile`? I thought the clarification was nice.

File: plugins/auth-backend/src/providers/microsoft/provider.ts [hunk 2]
```diff
@@ -205,32 +190,33 @@ export class MicrosoftAuthProvider implements OAuthHandlers {
...
+export type MicrosoftProviderOptions = {};
+
+export const createMicrosoftProvider = (
+  _options?: MicrosoftProviderOptions,
+): AuthProviderFactory => {
+  return ({ providerId, globalConfig, config, tokenIssuer }) =>
+    OAuthEnvironmentHandler.mapConfig(config, envConfig => {
+      const clientId = envConfig.getString('clientId');
+      const clientSecret = envConfig.getString('clientSecret');
+      const tenantID = envConfig.getString('tenantId');
```
Comment (freben) [inline plugins/auth-backend/src/providers/microsoft/provider.ts:202]:
```suggestion
      const tenantId = envConfig.getString('tenantId');
```

**Review by freben — APPROVED** [review]
Soooo nice that you got this ball rolling! This is definitely clearer

File: plugins/auth-backend/src/providers/auth0/provider.ts (hunk 1 above)
Comment (Rugvip) [reply plugins/auth-backend/src/providers/auth0/provider.ts:65]: I thought `rawProfile` didn't end up matching in the new context were it's actually being passed on to resolve the users. It's not `rawProfile -> profile` and then consume `profile` anymore, you end up needing both, with the `profile` just being the simplified version for the frontend settings page.

`passportProfile` could make sense too though, not sure if I think that's better than `fullProfile` though tbh.

File: plugins/auth-backend/src/providers/microsoft/provider.ts (hunk 2 above)
Comment (Rugvip) [reply plugins/auth-backend/src/providers/microsoft/provider.ts:202]: I mean sure, because I need to fix other stuff, but really not part of this PR xP

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4531-t1xgxaaebp3rdur12d5jxfrf73luuly1.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/602a610065f8172627f09bf1

**Pull request merged**